libs/pilight/core/common.c

```diff
@@ -1019,3 +1024,161 @@ int file_get_contents(char *file, char **content) {
+ * @return int Values < 0 indicate error. Values >= 0 indicate success.
+ * Where -1: syntax error, -2: DNS check for a domain failed, -8: severe internal error.
+ */
+int check_email_addr(const char *addr, int allow_lists, int check_domain_can_mail) {
+	if(addr == NULL) {
+		return -1;
+	}
+#if use_regexp
+	const regex_t *RE = NULL;
+	int reti = 0;
+
+	static const char email_pattern[] =
+		"^[" email_local_part_valid_chars "]+"
+		"@("
+			   "([A-Za-z0-9][-A-Za-z0-9]*[A-Za-z0-9])"
+			"(\\.[A-Za-z0-9][-A-Za-z0-9]*[A-Za-z0-9])+"
+		"|"
+			"[0-9]{1,3}(\\.[0-9]{1,3}){1,3}"	// IPv4 addr
+		")$";
+
+	if((RE = check_email_compiled_RE) == NULL) { // first check without lock.
+		// RE not compiled yet.
+		atomiclock();
+		if((RE = check_email_compiled_RE) == NULL) { // locked now, need to check again.
+			reti = regcomp(&check_email_common_RE, email_pattern, REG_EXTENDED);
+			if(reti == REG_NOERROR) {
+				check_email_compiled_RE = RE = &check_email_common_RE;
+				atexit(check_email_free_RE);
+			} else {
+				char errmsg[regerror(reti, &check_email_common_RE, NULL, 0)];
+				regerror(reti, &check_email_common_RE, errmsg, sizeof(errmsg));
+				regfree(&check_email_common_RE);
+				atomicunlock();	// do not logprintf while locked.
+				logprintf(LOG_ERR, "could not compile regex pattern '%s' to check email addresses: %s",
+						email_pattern, errmsg);
+				return -1;
+			}
+		}
+		atomicunlock();
+	}
+#else
+	static const char local_valid_chars[] = email_local_part_valid_chars;
+	static const char domain_valid_chars[] = AtoZ_atoz_0to9 ".-"; // . and - are not allowed everywhere.
+#endif
+
+	char _addr[strlen(addr)+1];
+	const char *at = NULL, *domain_name = NULL;
+	char *comma = NULL, *next_addr = NULL;
+
+	next_addr = strcpy(_addr, addr);	// get a writable copy.
+	while(next_addr != NULL) {
+		next_addr += strspn(next_addr, white_spaces);	// trim left
+		addr = next_addr;
+		comma = strchr(addr, ',');
+		if(comma != NULL) {
+			if (allow_lists == 0) {
+				return -1;	// comma seen but no lists allowed.
+			}
+			next_addr = comma + 1;
+		} else {
+			comma = next_addr + strlen(next_addr);
+			next_addr = NULL;
+		}
+		// trim right (comma points to ',' or term null):
+		while(comma > addr && strchr(white_spaces, comma[-1]) != NULL) {
+			comma--;
+		}
+		*comma = '\0';
+
+		// Now check addr. First some quick pre-checks:
+		at = strchr(addr, '@');
+		if(at == NULL || strstr(addr, "..") != NULL) {
+			return -1;	// no @ found or two consecutive dots found somewhere.
+		}
+		domain_name = at+1;
+#if use_regexp
+		reti = regexec(RE, addr, 0, NULL, 0);
+		if(reti == REG_NOMATCH) {
+			return -1;	// no error but no match.
+		}
+		if(reti != REG_NOERROR) {
+			char errmsg[regerror(reti, RE, NULL, 0)];
+			regerror(reti, RE, errmsg, sizeof(errmsg));
+			logprintf(LOG_ERR, "could not execute regex to check email addresses: %s", errmsg);
+			return -8;
+		}
+#else
+		// Do some checks without RE engine (catch many but not all wrong syntax):
```

Comment:
Isn't it possible to use a totally regex free system that works on both windows, linux, and BSD? You are implementing it already so better not do it redundant.